lib/booking_system.rb

```diff
@@ -0,0 +1,131 @@
+require_relative 'reservation'
+require_relative 'block_room'
+require_relative 'date_range'
+
+module Hotel
+  class BookingSystem
+
+    attr_reader :rooms, :reservations, :block_rooms
+
+    def initialize
+      @rooms = (1..20).to_a
+      @reservations = []
+      @block_rooms = []
+    end
+
+    def list_rooms
+      return @rooms
+    end
+
+    def list_reservations(date_range)
+      return @reservations.select { |res| res.date_range == date_range }
+    end
+
+    def list_available_rooms(date_range)
+      return @rooms if @reservations.empty?
+      return find_available_rooms(date_range)
+    end
+
+    def find_available_rooms(date_range)
+      return @rooms if @reservations.empty?
+      unavailable_rooms = @reservations.map { |res| res.room_number if date_range.dates_overlap?(res.date_range) }
+
+      unavailable_rooms_block = @block_rooms.map { |res| res.block_rooms if date_range.dates_overlap?(res.date_range)}.flatten
```

Comment:
This code has some problems. On line 31, you'll end up with a bunch of `nil`s in your array, since `map` must always produce a value even if none of the code in the block executes.

On line 33, not only will you have `nil`, but the elements in your array are more arrays, since that's what a `BlockRoom` stores. That means that after line 35 has run, you'll have an array like:

```ruby
[nil, 3, nil, nil, 5, nil, [8, 9, 10], nil]
```

A side effect of this is that your code does not respect blocks when making reservations!

You could use a combination of `Array#flatten` and `Array#compact` to fix it, or it could be that `map` isn't a great choice for this problem.

Reply:
I do love the way this code flows though - logic problems aside, this is an excellent example of loose coupling. Rather than pulling the dates out of the reservation and doing a bunch of math here, you call a method and let the date range take care of it. That makes this code easier to read, and it means if something about how overlaps are calculated changes this code won't have to. Good work!